### PR TITLE
Adding changes for ipsec-vpn

### PR DESCRIPTION
**Pull request by UtsavRaychaudhuri**

### CODE REVIEW

**Comment by UtsavRaychaudhuri** [comment]
@hkantare Please review the above pr

File: ibm/resource_ibm_ipsec_vpn.go [hunk 1]
```diff
@@ -0,0 +1,466 @@
...
+			},
+			"customer_peer_ip": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"internal_subnet_id": {
+				Type:     schema.TypeInt,
+				Optional: true,
+			},
+			"remote_subnet_id": { //customer subnet id . need atleast one customer subnet id for applying the configuratons
```
Comment (hkantare) [inline ibm/resource_ibm_ipsec_vpn.go]: As per description user can either provide remote_subnet_id or remote_subnet so add ConflictsWith: []string{"remote_subnet"}, respectively to both attributes..

File: ibm/resource_ibm_ipsec_vpn.go [hunk 2]
```diff
@@ -0,0 +1,466 @@
...
+	vpnID, _ := strconv.Atoi(d.Id())
+
+	vpn, err := services.GetNetworkTunnelModuleContextService(sess).
+		Id(vpnID).Mask(ipsecMask).
+		GetObject()
+	if err != nil {
+		return fmt.Errorf("Error retrieving firewall information: %s", err)
+	}
+	d.Set("name", *vpn.Name)
+	d.Set("internal_peer_ip_address", *vpn.InternalPeerIpAddress)
```
Comment (hkantare) [inline ibm/resource_ibm_ipsec_vpn.go]: In Read we should set back values of all attributes it helps during import scenario. Similar to InternalPerrIPAddress set back all attributes remote_subnet_id, phase_one, Phase_two....so on

File: ibm/resource_ibm_ipsec_vpn.go [hunk 3]
```diff
@@ -0,0 +1,466 @@
...
+	sess := meta.(ClientSession).SoftLayerSession()
+	vpnService := services.GetNetworkTunnelModuleContextService(sess)
+
+	vpnID, _ := strconv.Atoi(d.Id())
+
+	// Get billing item associated with the firewall
+	billingItem, err := vpnService.Id(vpnID).GetBillingItem()
+
+	if err != nil {
+		return fmt.Errorf("Error while looking up billing item associated with the firewall: %s", err)
```
Comment (hkantare) [inline ibm/resource_ibm_ipsec_vpn.go]: Log stmt needs to be corrected from firewall to ipsecvpn

File: ibm/resource_ibm_ipsec_vpn.go [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+			keylife := value["keylife"].(int)
+			vpn.PhaseTwoKeylife = &keylife
+		}
+	}
+	if d.HasChange("preshared_key") {
+		presharedkey := d.Get("preshared_key").(string)
+		vpn.PresharedKey = &presharedkey
+	}
+	if _, ok := d.GetOk("customer_peer_ip"); ok {
+		if d.HasChange("customer_peer_ip") {
```
Comment (hkantare) [inline ibm/resource_ibm_ipsec_vpn.go]: if and else loop has similar code...Do we need loop here?

File: ibm/resource_ibm_ipsec_vpn.go (hunk 2 above)
Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Some attributes are not set during creation. For eg:- if you mention only the datacenter name while creating an IPSecVPN only datacenter,id,name,phase_one and phase_two values are set. Others are not set.These attributes can only be set if the user specifically sets it while creating. So for now i am checking if these values are set and then setting them accordingly. Please tell me if u have a different approach in mind? I will wait for your reply before pushing the code.

Comment (hkantare) [reply ibm/resource_ibm_ipsec_vpn.go]: yes ..We need to check if values are set then only we need to set the attributes accordingly..

File: ibm/resource_ibm_ipsec_vpn.go [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+			"internal_peer_ip_address": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"phase_one": {
+				Type:     schema.TypeList,
```
[the remark this answers is not in the document]
Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Done

File: ibm/resource_ibm_ipsec_vpn.go (hunk 1 above)
Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Done

File: ibm/resource_ibm_ipsec_vpn.go (hunk 2 above)
Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Done. Please review

File: ibm/resource_ibm_ipsec_vpn.go (hunk 4 above)
Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Done. Please review

File: ibm/resource_ibm_ipsec_vpn.go (hunk 3 above)
Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Done. Please review

File: ibm/resource_ibm_ipsec_vpn.go [hunk 6]
```diff
@@ -0,0 +1,537 @@
...
+		GetObject()
+	if err != nil {
+		return fmt.Errorf("Error retrieving firewall information: %s", err)
+	}
+	d.Set("name", *vpn.Name)
+	d.Set("internal_peer_ip_address", *vpn.InternalPeerIpAddress)
+	if vpn.Datacenter != nil {
+		d.Set("datacenter", *vpn.Datacenter.Name)
+	}
+	phaseoneAttributesMap := make([]map[string]interface{}, 0, 1)
```
Comment (hkantare) [inline ibm/resource_ibm_ipsec_vpn.go]: Changes look fine...small chnage Can you move these attributesMap to respective flatten methods in structures.go (flattenIPSecPhaseOne,flattenIPSecPhaseTwo, flattenIPSecAddressTranslation, flattenIPSecCustomerSubnets)

Comment (UtsavRaychaudhuri) [reply ibm/resource_ibm_ipsec_vpn.go]: Done. Please review

**Comment by UtsavRaychaudhuri** [comment]
@hkantare Please review the above pr.

**Pull request merged**